rfd/0180-sso-as-mfa-method.md

```diff
@@ -0,0 +1,767 @@
+
+* Engineering: @rosstimothy || @codingllama
+* Product: @xinding33 || @klizhentas
+
+## What
+
+Provide the ability to satisfy Teleport MFA checks via a 3rd party identity
+provider with SSO.
+
+Per-session MFA must be supported for all Teleport clients, while additional
+features will be included as a stretch goal, such as MFA for admin actions and
+moderated sessions. Ideally we will support SSO as a first class MFA method.
+
+## Why
+
+Delegating MFA checks to a registered IdP has several potential benefits:
+- Administrators gain the ability to configure and monitor all authentication
+directly through an IdP.
+- Teleport can integrate with custom MFA mechanisms and devices registered
+setup through an IdP.
+- Improves UX for new SSO users by providing an MFA flow without requiring the
+user to manually register a new device.
+- Improve UX for existing users by providing multiple paths during MFA checks,
+such as when they don't have their WebAuthn device plugged in.
+- Improve security of new SSO users by requiring an SSO MFA check to add their
+first MFA device.
+
+## Details
+
+### UX
+
+This feature will provide users with additional paths towards passing MFA
+checks. In some cases, such as when the user has no MFA device registered,
+this will improve UX.
+
+Adding an SSO MFA option also comes with potential downsides:
+- Providing too many options without clarity (OTP, Webauthn, and SSO).
+- Locking users into SSO MFA in cases where WebAuthn may be preferred.
+- Automatically opening browser windows when WebAuthn may be preferred.
+
+These concerns will be addressed in the sections below.
+
+#### Devices
+
+When an auth connector has MFA enabled, Teleport will treat it as if any user
+originating from that auth connector has a corresponding SSO MFA device
+registered. This way, users can view their true MFA options.
+
+```console
+> tsh mfa ls
+Name     Type     Added at Last used                     
+-------- -------- -------- ---------
+yubi     WebAuthn ...      ...
+okta     SAML     ...      ...
+```
+
+Note: Teleport does not actually store a MFA device backend resource for each
+user. This means that interaction, e.g. `tsh mfa add/rm`, will not work and should
+result in a readable error message. If we want to support `tsh mfa add/rm`, we
+can implement the functionality behind the scenes with a new user preferences
+field, e.g. `use_sso_mfa`. I'm leaving this out of scope for now.
+
+#### Modes
+
+Auth connectors with MFA enabled will be configured with one of the following
+modes:
+- `optional`: SSO MFA is available to the user, but WebAuthn is preferred when
+  applicable.
+- `preferred`: SSO MFA is preferred over WebAuthn, but Webauthn is still
+  available to the user.
+- `required`: SSO MFA is the only available MFA method.
+
+Each mode has UX implications that will be explored in more detail below.
+
+Note: Like WebAuthn, SSO will always be preferred over OTP, unless OTP is
+requested with `tsh --mfa-mode=otp`.
```

Comment:
I see how it could be confusing, but I'm more concerned with the configuration being foolproof. Confusion can be solved with documentation, but misconfigurations can lead to real problems.

One way or another, MFA settings need to be set in the Auth connector to handle things like the OIDC client id/secret and SAML entity descriptor. Think about how things could go wrong if we move some settings to the auth preference:

Ex 1: Add `cap.second_factor: sso` to require SSO MFA
- What about non-sso users? is webauthn required? totp allowed? etc.
- What if SSO MFA is required but a user's auth connector isn't configured for MFA? Can they still use webauthn?

Ex2: Add `cap.sso_mfa_mode` with same values: optional, preferred, and required
- Same issues with `required` as `cap_second_factor: sso` from Ex1
- What if a cluster has 2 Auth connectors. One IdP integrates with a custom MFA device which is preferred over webauthn. In this case, the admin will set `required`. But the other IdP only integrates with WebAuthn, so using WebAuthn natively with teleport would be much better UX, and `optional` would be a better setting. This must then be set in the auth connector, or at least have an overwrite option in the auth connector?

Reply:
Maybe we could include something in the CAP's status about the SSO MFA configuration, if it's set, so at least it's visible in the resource?

Reply:
Removed mfa modes (just prefer WebAuthn > SSO > OTP) and added `cap.sso.enabled` - efdb76ecb4aec705db441e13bf21d18e31cf09ef

Reply:
Great idea, I'll update the RFD with this.

> We could do the list in order of preference too, so [webauthn, sso] favors webauthn, while [sso, webauthn] favors sso.

Adding a note but I think this isn't necessary for now.